qualtran/bloqs/state_preparation/state_preparation_alias_sampling.py

```diff
@@ -235,6 +235,219 @@ def build_call_graph(self, ssa: 'SympySymbolAllocator') -> Set['BloqCountT']:
+    $$
+
+    where $\mathrm{ind}_l$ is the index of the $l$-th non-zero coefficient,
+    and the probabilities $p_\ell$ are $\mu$-bit binary approximations to the true values and
+    where the temporary register must be treated with care, see the details in Section 5 of
+    reference [1] and Section III.D. of the reference [2].
+
+    The preparation is equivalent to [classical alias sampling]
+    (https://en.wikipedia.org/wiki/Alias_method): we sample `l` with probability `p[l]` by first
+    selecting `l` uniformly at random and then returning `ind[l]` with probability `keep[l] / 2**mu`;
+    otherwise returning `alt[l]`.
+    This bloq is nearly identical to :class:`StatePreparationByAliasSampling`, except that this loads
+    the non-zero coefficient indices as well from the QROM.
+
+    Signature:
+        selection: The input/output register $|\mathrm{ind}_l\rangle$ of size lg(L) where the desired
+            coefficient state is prepared.
+        temp: Work space comprised of sub signature:
+            - sigma: A mu-sized register containing uniform probabilities for comparison against
+                `keep`.
+            - alt: A lg(L)-sized register of alternate indices
+            - keep: a mu-sized register of probabilities of keeping the initially sampled index.
+            - one bit for the result of the comparison.
+
+    This gate corresponds to the following operations:
+     - UNIFORM_L on the sparse_index register
+     - H^mu on the sigma register
+     - QROM addressed by the selection register into the alt and keep signature.
+     - LessThanEqualGate comparing the keep and sigma signature.
+     - Coherent swap between the selection register and alt register if the comparison
+       returns True.
+
+    Total space will be (2 * log(L) + 2 mu + 1) work qubits + log(L) ancillas for QROM.
+    The 1 ancilla in work qubits is for the `LessThanEqualGate` followed by coherent swap.
+
+    References:
+        [1] [Qubitization of Arbitrary Basis Quantum Chemistry Leveraging Sparsity and Low Rank Factorization](https://arxiv.org/pdf/1902.02134#page=15.30)
+        Berry et al. (2019). Section 5, Eqs. 43, 44.
+        [2] [Encoding Electronic Spectra in Quantum Circuits with Linear T Complexity](https://arxiv.org/abs/1805.03662).
+        Babbush et al. (2018). Section III.D. and Figure 11.
+    """
+    selection_registers: Tuple[Register, ...] = attrs.field(
+        converter=lambda v: (v,) if isinstance(v, Register) else tuple(v)
+    )
+    index: Union[Shaped, NDArray[np.int_]] = attrs.field(eq=_data_or_shape_to_tuple)
+    alt: Union[Shaped, NDArray[np.int_]] = attrs.field(eq=_data_or_shape_to_tuple)
+    keep: Union[Shaped, NDArray[np.int_]] = attrs.field(eq=_data_or_shape_to_tuple)
+    mu: SymbolicInt
+    sum_of_lcu_coeffs: SymbolicFloat
+
+    @classmethod
+    def from_lcu_probs(
+        cls,
+        lcu_probabilities: Sequence[float],
```

Comment:
Should we change the input format and expect a dictionary `Dict[int, float]` that specifies the coefficients for the sparse indices? We can construct a dense `lcu_probabilities` as part of the function?

Reply:
yeah this makes more sense, I added this as the default method